### PR TITLE
Expression printer to make debugging of expressions easier - formats the expression to a more readable form than the "Debugger View"

### PR DESCRIPTION
**Pull request by maumar**
Also added a few classes that format the value of a constant expression, depending on it's type, e.g reflecting on private fields of a CommandBuilder to display the actual SQL that will be generated.
This is a debug-only feature to prevent the performance hit.

### CODE REVIEW

**Comment by maumar** [comment]
@anpete

**Comment by maumar** [comment]
rebased on the latest rename

File: src/EntityFramework.Relational/Query/Internal/CommandBuilderPrinter.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        public override bool TryPrintConstant(object value, IndentedStringBuilder stringBuilder)
+        {
+            var commandBuilder = value as CommandBuilder;
+            if (commandBuilder != null)
+            {
+                stringBuilder.AppendLine("SelectExpression: ");
+                stringBuilder.IncrementIndent();
+
+                var sqlGeneratorFactoryInfo = commandBuilder.GetType().GetField("_sqlGeneratorFactory", BindingFlags.NonPublic | BindingFlags.Instance);
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/Internal/CommandBuilderPrinter.cs]: Don't use reflection here or below.

File: src/EntityFramework.Core/Query/EntityQueryModelVisitor.cs [hunk 2]
```diff
@@ -106,6 +107,15 @@ protected virtual ExpressionVisitor CreateOrderingExpressionVisitor([NotNull] Or
             }
         }
 
+        protected virtual List<IConstantPrinter> GetConstantPrinters()
```
Comment (anpete) [inline src/EntityFramework.Core/Query/EntityQueryModelVisitor.cs]: This should be part of the printer class.

File: src/EntityFramework.Core/Query/Internal/CollectionConstantPrinter.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.Data.Entity.Internal;
+
+namespace Microsoft.Data.Entity.Query.Internal
+{
+    public class CollectionConstantPrinter : ConstantPrinterBase, IConstantPrinter
```
Comment (anpete) [inline src/EntityFramework.Core/Query/Internal/CollectionConstantPrinter.cs]: Can we make these internal to the top-level printer class?

File: src/EntityFramework.Core/Query/Internal/ExpressionPrinter.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using Microsoft.Data.Entity.Internal;
+
+namespace Microsoft.Data.Entity.Query.Internal
+{
+    public class ExpressionPrinter
+    {
+        public string Print(Expression expression, List<IConstantPrinter> constantPrinters)
```
Comment (anpete) [inline src/EntityFramework.Core/Query/Internal/ExpressionPrinter.cs]: Merge this into the visitor below.

**Comment by maumar** [comment]
@anpete  updated

File: src/EntityFramework.Core/Query/Internal/ExpressionPrinter.cs [hunk 5]
```diff
@@ -0,0 +1,533 @@
...
+
+namespace Microsoft.Data.Entity.Query.Internal
+{
+    public class ExpressionPrinter : ExpressionVisitor, IExpressionPrinter
+    {
+        private IndentedStringBuilder _stringBuilder;
+        private List<IConstantPrinter> _constantPrinters;
+        private Dictionary<ParameterExpression, string> _parametersInScope;
+
+        private readonly Action<IndentedStringBuilder, string> _append = (sb, s) => sb.Append(s);
```
Comment (anpete) [inline src/EntityFramework.Core/Query/Internal/ExpressionPrinter.cs]: Duplicated above

File: test/EntityFramework.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs [hunk 6]
```diff
@@ -78,8 +129,90 @@ var customers
...
+            element: prm1
+        )
+        , 
+        selector: (QueryResultScope prm1) => IEnumerable<QueryResultScope<Customer>> _Include(
+            queryContext: (RelationalQueryContext) prm0, 
+            innerResults: IEnumerable<QueryResultScope<Customer>> _ShapedQuery(
+                queryContext: prm0, 
+                commandBuilder: VALUE(CommandBuilder), 
+                shaper: (ValueBuffer prm2) => QueryResultScope<Customer> CreateEntity(
+                    querySource: VALUE(from Customer c in value(Microsoft.Data.Entity.Query.EntityQueryable`1[Microsoft.Data.Entity.FunctionalTests.TestModels.Northwind.Customer])),
```
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs]: Remove namespace?

File: test/EntityFramework.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs [hunk 7]
```diff
@@ -29,8 +29,59 @@ var customers
...
+                queryContext: prm0, 
+                parentQueryResultScope: prm1, 
+                valueBuffer: prm2, 
+                valueBufferOffset: VALUE(0), 
+                entityType: VALUE(Microsoft.Data.Entity.FunctionalTests.TestModels.Northwind.Customer), 
+                queryStateManager: VALUE(True), 
+                entityKeyFactory: VALUE(SimpleNullSentinelEntityKeyFactory`1), 
+                keyProperties: List<Property> 
+                { 
+                    Microsoft.Data.Entity.Metadata.Property,
```
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs]: ToString?

File: test/EntityFramework.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs [hunk 8]
```diff
@@ -29,8 +29,59 @@ var customers
...
+                keyProperties: List<Property> 
+                { 
+                    Microsoft.Data.Entity.Metadata.Property, 
+                }
+                , 
+                materializer: (ValueBuffer prm3) => 
+                {
+                    var var4
+                    var4 = new Customer()
+                    var4.CustomerID = (string) object prm3.get_Item(VALUE(0))
```
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs]: "VALUE" just seems to add noise here...

File: test/EntityFramework.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs [hunk 9]
```diff
@@ -78,8 +129,90 @@ var customers
...
     Tracking query sources: [c]
-    Compiled query expression.
-", TestSqlLoggerFactory.Log);
+    TRACKED: True
+(QueryContext prm0, QueryResultScope prm1) => () => IEnumerable<Customer> _Select(
+    source: IEnumerable<QueryResultScope> _SelectMany(
+        source: IEnumerable<QueryResultScope> _ToSequence(
+            element: prm1
+        )
+        ,
```
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs]: Newline before ","

File: src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 10]
```diff
@@ -895,7 +902,12 @@ protected override Expression VisitParameter(ParameterExpression parameterExpres
 
             if (_commandParameters.All(commandParameter => commandParameter.Name != parameterName))
             {
-                var value = _parameterValues[parameterExpression.Name];
+                object value;
+                if (!_parameterValues.TryGetValue(parameterExpression.Name, out value))
+                {
+                    value = "dummyPrm";
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: string.Empty as a small optimization?

File: src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 11]
```diff
@@ -492,10 +495,14 @@ protected virtual Expression VisitNotIn(InExpression inExpression)
                 }
 
                 var inParameter = inValue as ParameterExpression;
-                if (inParameter != null
-                    && _parameterValues[inParameter.Name] != null)
+                if (inParameter != null)
                 {
-                    inValuesNotNull.Add(inValue);
+                    object parameterValue;
+                    _parameterValues.TryGetValue(inParameter.Name, out parameterValue);
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: Not using return value?

File: src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 12]
```diff
@@ -459,7 +461,8 @@ protected virtual Expression VisitNotIn(InExpression inExpression)
                 var inParameter = inValue as ParameterExpression;
                 if (inParameter != null)
                 {
-                    var parameterValue = _parameterValues[inParameter.Name];
+                    object parameterValue;
+                    _parameterValues.TryGetValue(inParameter.Name, out parameterValue);
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: Not using return value?

File: src/EntityFramework.Core/EntityFramework.Core.csproj [hunk 13]
```diff
@@ -226,8 +226,12 @@
     <Compile Include="Query\ExpressionVisitors\ParameterExtractingExpressionVisitor.cs" />
     <Compile Include="Query\ExpressionVisitors\QueryAnnotatingExpressionVisitor.cs" />
     <Compile Include="Query\ExpressionVisitors\ReducingExpressionVisitor.cs" />
+    <Compile Include="Query\Internal\ConstantExpressionPrinterBase.cs" />
+    <Compile Include="Query\Internal\ExpressionPrinter.cs" />
+    <Compile Include="Query\Internal\IConstantExpressionPrinter.cs" />
     <Compile Include="Query\IEntityQueryProvider.cs" />
     <Compile Include="Query\IIncludableQueryable.cs" />
+    <Compile Include="Query\Internal\IExpressionPrinter.cs" />
```
Comment (anpete) [inline src/EntityFramework.Core/EntityFramework.Core.csproj]: Do we need these new interfaces? Can ConstantExpressionPrinterBase be nested?

Comment (maumar) [inline src/EntityFramework.Core/EntityFramework.Core.csproj]: done

File: src/EntityFramework.Core/Query/Internal/ExpressionPrinter.cs (hunk 5 above)
Comment (maumar) [inline src/EntityFramework.Core/Query/Internal/ExpressionPrinter.cs]: removed funcs from ConstantPrinterBase, added static properties to ExpressionPrinter instead, that other classes use as well

File: src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs (hunk 12 above)
Comment (maumar) [inline src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: this code just flattens the parameters which are collections, parameterValue is used for this purpose in line 466

File: src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs (hunk 11 above)
Comment (maumar) [inline src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: Using it in a line below (502), this code only filters out the parameters which are null

File: src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs (hunk 10 above)
Comment (maumar) [inline src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: done

File: test/EntityFramework.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs (hunk 8 above)
Comment (maumar) [inline test/EntityFramework.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs]: removed across the board

File: test/EntityFramework.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs (hunk 7 above)
Comment (maumar) [inline test/EntityFramework.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs]: This is a problem with DI setup, RelationalExpresionPrinter should already handle this

File: src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs (hunk 11 above)
Comment (anpete) [inline src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: Return value of TryGetValue - would be typical to conditionalize on that.

**Comment by maumar** [comment]
@anpete updated

**Comment by anpete** [comment]
:shipit:

**Pull request closed without merging**